hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQLDialect.java

```diff
@@ -465,20 +465,23 @@ public long getFractionalSecondPrecisionInNanos() {
 
 	@Override
 	public String timestampaddPattern(TemporalUnit unit, TemporalType temporalType, IntervalType intervalType) {
-		if ( intervalType != null ) {
-			return "(?2+?3)";
-		}
-		switch ( unit ) {
+		return intervalType != null
+				? "(?2+?3)"
+				: "cast(?3+" + intervalPattern( unit ) + " as " + temporalType.name().toLowerCase() + ")";
```

Comment:
Not sure what happens here exactly or why this is necessary, but it would be awesome if we could avoid the explicit cast here if possible. Do you think there is a way to do that i.e. by extending the `TimestampaddFunction` for PostgreSQL and render the cast around based on whatever the condition is that makes implicit typing fail?

Reply:
I just quickly checked and it seems like what PostgreSQL does is "reasonable". `current_date + interval '2 days'` always produces a `timestamp`, because how could it know if the interval contains a time part or not?
The second form `date - date` is special, as noted in the documentation, since it produces the amount of seconds between two dates instead.

I hope our type system produces a `timestamp` as well for `date + duration`? I handled these sort of special cases for timestampdiff in DB2 as well.

Reply:
I mean I don't know what postgres you tried this on, but on 15 it's easy to see the problem:

```
hreact=# select current_date - (current_date + interval '2 days');
 ?column? 
----------
 -2 days
(1 row)

hreact=# select current_date - date '2022-12-1';
 ?column? 
----------
      263
(1 row)
```

These are not the same type:

```
hreact=# select pg_typeof(current_date - date '2022-12-1');
 pg_typeof 
-----------
 integer
(1 row)

hreact=# select pg_typeof(current_date - (current_date + interval '2 days'));
 pg_typeof 
-----------
 interval
(1 row)
```

Again, you can easily see how this goes wrong by making the change you suggest and running the test.